src/PerfCounters.cc

```diff
@@ -32,23 +32,29 @@ namespace rr {
 #define PERF_COUNT_RR 0x72727272L
 
 static bool attributes_initialized;
-// pmu_bug_flags is an architecture dependent flags to determine
-// what bugs need to be checked.
-// Current, this is simply the uarch on x86 and unused on aarch64.
-static int pmu_bug_flags;
 // At some point we might support multiple kinds of ticks for the same CPU arch.
 // At that point this will need to become more complicated.
-static struct perf_event_attr ticks_attr;
-static struct perf_event_attr minus_ticks_attr;
-static struct perf_event_attr cycles_attr;
-static struct perf_event_attr llsc_fail_attr;
-static uint32_t pmu_flags;
-static uint32_t skid_size;
+struct perf_event_attrs {
+  // bug_flags is an architecture dependent flags to determine
+  // what bugs need to be checked.
+  // Current, this is simply the uarch on x86 and unused on aarch64.
+  int bug_flags{0};
+  perf_event_attr ticks{};
+  perf_event_attr minus_ticks{};
+  perf_event_attr cycles{};
+  perf_event_attr llsc_fail{};
+  uint32_t pmu_flags{0};
+  uint32_t skid_size{0};
+};
+// If this contains more than one element, it's indexed by the CPU index.
+static std::vector<perf_event_attrs> perf_attrs;
+static uint32_t pmu_semantics_flags;
 
 static bool pmu_checked;
 static bool has_ioc_period_bug;
 static bool only_one_counter;
 static bool activate_useless_counter;
+static int pmu_attr_index;
```

Comment:
Can you document what this means?

Reply:
I think this probably shouldn't exist and instead we can just select the right PMU attributes entry based on `PerfCounter::cpu_binding_`?

Reply:
This is used by the static function `PerfCounters::skid_size()`

Reply:
Let's make `skid_size()` nonstatic.

Reply:
There are actually somewhat more than that. If we are removing this, we should also remove the assumption in this file that a single PMU type is used at runtime. I don't think it makes much sense otherwise to remove this global state.

For that to happen, I think the two more things that needs to happen,
* The other global states should be moved into the array. This should be relatively easy to do.
* The cpu_binding member is actually not always initialized. This happens for the diversion task (and for which I think I've left a comment somewhere).

Reply:
Yes, I think we should do these things.